Add support for S1G beacon to the dot11 layer. (#2)

rkinder2023 opened the pull request:
Add support for 802.11ah (S1G) beacon parsing.

 - Includes unit test parsing and confirming an S1G beacon
 - Includes support for the new Frame Control format for type=3, subtype=1 (S1G beacon)
 - Includes changes to support addressing used in S1G beacon.

All dot11 unit tests pass with this change.

Fixes #4439

rkinder2023 commented:
I'm really not sure what to do here as the commit into my master branch shows coverage at the correct level (80.73%). I don't know why coverage is reported so low here.

https://app.codecov.io/github/rkinder2023/scapy-s1g/commit/d377b137fa67845896442fd01615abcd883bc590

Any hints from the maintainers?

rkinder2023 commented:
G'day folks, can I please get approval for the workflow (review would be great too :)? @gpotter2, @guedou, @p-l- ?

p-l- commented on scapy/layers/dot11.py:
```diff
@@ -712,15 +720,15 @@ class Dot11(Packet):
         ConditionalField(
             _Dot11MacField("addr2", ETHER_ANY, 2),
-            lambda pkt: (pkt.type != 1 or
+            lambda pkt: (pkt.type not in [1, 3] or
```
```suggestion
            lambda pkt: (pkt.type not in {1, 3} or
```

p-l- commented on scapy/layers/dot11.py:
```diff
@@ -712,15 +720,15 @@ class Dot11(Packet):
                          ((pkt.type, pkt.subtype) == (1, 6) and pkt.cfe == 6)),
         ),
-        ConditionalField(LEShortField("SC", 0), lambda pkt: pkt.type != 1),
+        ConditionalField(LEShortField("SC", 0), lambda pkt: pkt.type not in [1, 3]),
```
```suggestion
        ConditionalField(LEShortField("SC", 0), lambda pkt: pkt.type not in {1, 3}),
```

p-l- commented on scapy/layers/dot11.py:
```diff
@@ -702,6 +702,14 @@ class Dot11(Packet):
+                (
+                    # FIXME: bw1/2/3 are a 3-bit field.
+                    # Unsure how to do this properly.
+                    FlagsField("FCfield", 0, 8,
+                               ["next_tbtt", "comp_ssid", "ano", "bw1",
+                                "bw2", "bw3", "security", "AP_PM"]),
```
Good question; apart from doing something like that, I don't know (and that's not really great, so that's not a suggestion)...

@gpotter2 @guedou do you have a better idea?

```suggestion
                    FlagsField("FCfield2", 0, 2, ["security", "AP_PM"]),
                    BitField("FCfield_bw", 0, 3),
                    FlagsField("FCfield", 0, 3, ["next_tbtt", "comp_ssid", "ano"]),
```

gpotter2 replied:
I think in that case I would split in in BitFields directly (so in this case 6 fields). That's what's usually done in Scapy (as opposed to FlagsFields)

rkinder2023 replied:
I'll try BitFields directly. Thanks @gpotter2 and @p-l- for your reviews.

rkinder2023 replied:
New change pushed which uses a combo of BitField and FlagsField. I am not willing to modify the existing FCField to be all BitFields as it may cause significant breakage of user scripts.

rkinder2023 commented:
Could I please get another review? I have modified to use a set on the 'type in' as per @p-l-, and updated the frame control field to split into three sections for type 3, subtype 1 (@p-l-, @gpotter2 suggestions). This seems the best solution as a change to the underlying frame control to use BitField for all types and subtypes will introduce significant back-compat issues.

Let me know if there is any more testing/unit testing you'd like on this change? I'd like to push this one in as I have more changes to support the S1G IEs within the beacon queued up.

Thanks guys!

rkinder2023 commented:
Hi folks, any further review or comments? @p-l- @gpotter2 @guedou?

gpotter2 commented:
Yeah sorry for the delay. I wanted to test your PR, and still couldn't get around to do it.

rkinder2023 commented:
@gpotter2, PCAP parsing for the trace I have doesn't work, I'm working on an acceptable solution to this. I managed to hack something in place in order to extract this beacon frame, but don't yet have something ready for review.

rkinder2023 commented:
[S1G-beacon.pcap.gz](https://github.com/user-attachments/files/16593257/S1G-beacon.pcap.gz)

This is the S1G beacon I used for developing the patch. The hack I used to get the pcap to load is here:

https://github.com/rkinder2023/scapy-s1g/commit/fda0baba2845288fc1b1b8d54489807c32650c39

I'm sure there's a better way to make TLVs work properly, but I don't know what it is.

rkinder2023 commented:
@gpotter2, any thoughts on this? Anything else you need me to provide?

gpotter2 commented:
Thanks for the PR and terribly sorry for the wait.
